src/cryptography/hazmat/primitives/serialization/ssh.py

```diff
@@ -1035,3 +1039,353 @@ def serialize_ssh_public_key(public_key: _SSH_PUBLIC_KEY_TYPES) -> bytes:
+            critical_options=self._critical_options,
+            extensions=self._extensions,
+        )
+
+    def valid_principals(
+        self, valid_principals: typing.List[bytes]
+    ) -> "SSHCertificateBuilder":
+        if not all(isinstance(x, bytes) for x in valid_principals):
+            raise TypeError("principals must be a list of bytes")
+        if self._valid_principals is not None:
+            raise ValueError("valid_principals already set")
+
+        return SSHCertificateBuilder(
+            public_key=self._public_key,
+            serial=self._serial,
+            type=self._type,
+            key_id=self._key_id,
+            valid_principals=valid_principals,
+            valid_before=self._valid_before,
+            valid_after=self._valid_after,
+            critical_options=self._critical_options,
+            extensions=self._extensions,
+        )
+
+    def valid_before(
+        self, valid_before: datetime.datetime
+    ) -> "SSHCertificateBuilder":
+        if not isinstance(valid_before, datetime.datetime):
+            raise TypeError("valid_before must be a datetime")
+        if _datetime_to_utc_timestamp(valid_before) < 0:
+            raise ValueError("valid_before must be after the Unix epoch")
+        if self._valid_before is not None:
+            raise ValueError("valid_before already set")
+
+        return SSHCertificateBuilder(
+            public_key=self._public_key,
+            serial=self._serial,
+            type=self._type,
+            key_id=self._key_id,
+            valid_principals=self._valid_principals,
+            valid_before=valid_before,
+            valid_after=self._valid_after,
+            critical_options=self._critical_options,
+            extensions=self._extensions,
+        )
+
+    def valid_after(
+        self, valid_after: datetime.datetime
+    ) -> "SSHCertificateBuilder":
+        if not isinstance(valid_after, datetime.datetime):
+            raise TypeError("valid_after must be a datetime")
+        if _datetime_to_utc_timestamp(valid_after) < 0:
+            raise ValueError("valid_after must be after the Unix epoch")
+        if self._valid_after is not None:
+            raise ValueError("valid_after already set")
+
+        return SSHCertificateBuilder(
+            public_key=self._public_key,
+            serial=self._serial,
+            type=self._type,
+            key_id=self._key_id,
+            valid_principals=self._valid_principals,
+            valid_before=self._valid_before,
+            valid_after=valid_after,
+            critical_options=self._critical_options,
+            extensions=self._extensions,
+        )
+
+    def add_critical_option(
+        self, name: bytes, value: bytes
+    ) -> "SSHCertificateBuilder":
+        if not isinstance(name, bytes) or not isinstance(value, bytes):
+            raise TypeError("name and value must be bytes")
+        if name in [x[0] for x in self._critical_options]:
```

Comment:
```suggestion
        if name in [name for name, _ in self._critical_options]:
```

and add a comment that its O(n^2)